server/src/main/java/org/apache/cassandra/sidecar/cluster/InstancesMetadataImpl.java

```diff
@@ -64,37 +74,70 @@ public InstancesMetadataImpl(List<InstanceMetadata> instanceMetadataList, DnsRes
     }
 
     @Override
-    public InstanceMetadata instanceFromId(int id) throws NoSuchSidecarInstanceException
+    public InstanceMetadata instanceFromId(int id) throws NoSuchCassandraInstanceException
     {
         InstanceMetadata instanceMetadata = idToInstanceMetadata.get(id);
         if (instanceMetadata == null)
         {
-            throw new NoSuchSidecarInstanceException("Instance id '" + id + "' not found");
+            throw new NoSuchCassandraInstanceException("Instance id '" + id + "' not found");
         }
         return instanceMetadata;
     }
 
     @Override
-    public InstanceMetadata instanceFromHost(String host) throws NoSuchSidecarInstanceException
+    public InstanceMetadata instanceFromHost(String hostOrIpAddress) throws NoSuchCassandraInstanceException
     {
-        InstanceMetadata instanceMetadata = hostToInstanceMetadata.get(host);
+        // if the host string is hostname string, resolve the ip address and loop up again
+        InstanceMetadata instanceMetadata = hostNameToInstanceMetadata.get(hostOrIpAddress);
+        if (instanceMetadata == null)
+        {
+            // if the 'host' string is IP address string
+            instanceMetadata = ipToInstanceMetadata.get(hostOrIpAddress);
+        }
+
+        // maybe the IP of the hostname has been updated in the DNS, and client is using the updated IP
+        // update the ipToInstanceMetadata and look up with the IP address again
+        if (instanceMetadata == null)
+        {
+            updateIpToInstanceMetadata();
+            instanceMetadata = ipToInstanceMetadata.get(hostOrIpAddress);
+        }
+
         if (instanceMetadata == null)
         {
+            throw new NoSuchCassandraInstanceException("Instance with host address '" + hostOrIpAddress + "' not found");
+        }
+        return instanceMetadata;
```

Comment:
Can we return early here?
```suggestion
        // if the host string is hostname string, resolve the ip address and loop up again
        InstanceMetadata instanceMetadata = hostNameToInstanceMetadata.get(hostOrIpAddress);
        if (instanceMetadata != null)
        {
            return instanceMetadata;
        }

        // if the 'host' string is IP address string
        instanceMetadata = ipToInstanceMetadata.get(hostOrIpAddress);
        if (instanceMetadata != null)
        {
            return instanceMetadata;
        }

        updateIpToInstanceMetadata();
        instanceMetadata = ipToInstanceMetadata.get(hostOrIpAddress);
        if (instanceMetadata == null)
        {
            throw new NoSuchCassandraInstanceException("Instance with host address '" + hostOrIpAddress + "' not found");
        }
        return instanceMetadata;
```